Feature/local for details

itsahmedsalah opened the pull request.

itsahmedsalah committed:
Add the dependence for the movie local datasource

itsahmedsalah committed:
feat: Add local data source for movie cast

This commit introduces the local data source implementation for movie cast information using Room.

The following changes are included:
- Added `CastLocalDataSource` interface in `repository/movie`.
- Implemented `CastDataSourceImp` in `datasource/local/movie` which uses a `CastDao`.
- Created `CastDao` with methods to add and retrieve cast entities.
- Defined `CastEntity` for Room database.
- Added `CastsDto` and `CastDto` for data transfer.
- Updated Gradle dependencies for Room and repository modules.
- Removed unused `MyClass.kt` from `repository/movie`.

itsahmedsalah committed:
Convert design system PNGs to WebP format

itsahmedsalah committed:
Add Room database entities for media details

Mohamed-Elhanafy reviewed (changes requested):
Please fix the build error and update the release-info.text file

itsahmedsalah committed:
Made the LocalDataSources for the MediaDetails

itsahmedsalah committed:
Add doa and datasource implementation for movie and reviews and production company

itsahmedsalah committed:
Add Room database and DAOs for movie details

This commit introduces a Room database (`MovieDetailDataBase`) for storing movie-related data locally. It includes DAOs for:
- Movies (`MovieDao`)
- Cast (`CastDao`)
- Reviews (`ReviewDao`)
- Galleries (`GalleryDao`)

DataSources implementations for these DAOs are also added. The `MovieDao` has been updated to include insert and query methods.

Additionally, the build configuration for the `datasource:local:movie` module has been updated to an Android library.

itsahmedsalah committed:
Added the movie detail converter

itsahmedsalah committed:
Add DAOs for Country, Genre, Image, and ProductionCompany

itsahmedsalah committed:
Added Image, Genre, and Country LocalDataSources interface

itsahmedsalah committed:
Rename GenreDao to GenresDao and ImageDao methods

- Renamed `GenreDao` to `GenresDao`.
- Renamed `addImages` to `addImage` and `getImages` to `getImage` in `ImageDao`.
- Added new local data source implementations for:
    - Countries (`CountriesLocalDataSourceImpl`)
    - Genres (`GenresLocalDataSourceImpl`)
    - Images (`ImageLocalDataSourceImp`)
    - Production Companies (`ProductionCompanyLocalDataSource`)
- Added utility functions:
    - `getCurrentDate()` to get the current date and time.
    - `detectLanguage()` to get the default system language.
- Added a `Util` object containing a list of `CountryEntity` objects.

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/local-for-details

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/entity/CastEntity.kt:
```diff
@@ -0,0 +1,12 @@
+    val id: Int,
+    val name: String,
+    val imageUri: String,
+)
```
the @PrimaryKey is enough no need for primaryKeys = ["id"]

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/entity/CastEntity.kt:
```diff
@@ -0,0 +1,12 @@
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "Cast_Table", primaryKeys = ["id"])
```
How the cast is related to a movie ?

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/GalleryDao.kt:
```diff
@@ -0,0 +1,15 @@
+    suspend fun addGallery(gallery: List<GalleryEntity>)
+    @Query("SELECT * FROM gallery_table")
+    suspend fun getGallery(): List<GalleryEntity>
+}
```
why adding list of GalleryEntity?

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/entity/GalleryEntity.kt:
```diff
@@ -0,0 +1,11 @@
+    @PrimaryKey(autoGenerate = true)
+    val id: Int,
+    val images: List<ImageEntity>,
+)
```
how the GalleryEntity is related to a movie?

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/GenresDao.kt:
```diff
@@ -0,0 +1,18 @@
+    @Query("SELECT * FROM Genres_table")
+    suspend fun getGenres(): List<GenreEntity>
+
+}
```
you don't need dao for GenreEntity
you are saving it in the MovieEntity as a list

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/ImageDao.kt:
```diff
@@ -0,0 +1,16 @@
+
+    @Query("SELECT * FROM images_table")
+    suspend fun getImage(): ImageEntity
+}
```
you don't need Dao for ImageEntity you are saving it as a list in the Gallery

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/ProductionCompanyDao.kt:
```diff
@@ -0,0 +1,16 @@
+
+    @Query("SELECT * FROM productionCompany_table")
+    suspend fun getProductionCompanies(): List<ProductionCompanyEntity>
+}
```
you don't need ProductionCompanyDao  you are saving it as a list in the movie

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/entity/ReviewEntity.kt:
```diff
@@ -0,0 +1,17 @@
+    val avatarUrl: String,
+    val username: String,
+    val rating: Double,
+)
```
how the ReviewEntity is related to a movie?

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/CastDao.kt:
```diff
@@ -0,0 +1,15 @@
+@Dao
+interface CastDao {
+    @Insert
+    suspend fun addCast(genres: List<CastEntity>)
```
solve the typo

itsahmedsalah replied:
what do you mean?

Joseph-Sameh-0 reviewed (changes requested):
Please resolve the comments

itsahmedsalah committed:
Refactor: Modify MovieDao and LocalDataSource for single movie operations

itsahmedsalah committed:
Rename the package

itsahmedsalah committed:
Merge remote-tracking branch 'origin/develop' into feature/local-for-details

# Conflicts:
#	gradle/libs.versions.toml

itsahmedsalah committed:
Fix typo in package name for local data sources

itsahmedsalah committed:
Refactor: Consolidate movie detail entities into MovieEntity

- Removed separate Dao and LocalDataSource for Countries, Genres, and ProductionCompanies.
- Integrated `List<ProductionCompanyEntity>`, `List<GenreEntity>`, and `CountryEntity` directly into `MovieEntity` using Room TypeConverters.
- Updated `ReviewEntity` to include `movieId` and a foreign key constraint to `MovieEntity`.
- Modified `ReviewDao` to fetch reviews by `movieId`.
- Made `CountryEntity`, `GenreEntity`, and `ProductionCompanyEntity` serializable and removed `@PrimaryKey` as they are now embedded.

itsahmedsalah committed:
Refactor: Remove ImageDao and ImageLocalDataSource, integrate functionality into Gallery

itsahmedsalah committed:
Refactor: Remove ImageDao and ImageLocalDataSource, integrate functionality into Gallery

mohammed-akkad committed:
fix build gradle

mohammed-akkad committed:
update build gradle

mohammed-akkad committed:
update movie detail converter

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/CastDao.kt:
```diff
@@ -0,0 +1,16 @@
+@Dao
+interface CastDao {
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun addCast(genres: List<CastEntity>)
```
cast not genres

Joseph-Sameh-0 commented on datasource/local/movie/src/main/java/com/datasource/local/dao/MovieDao.kt:
```diff
@@ -0,0 +1,16 @@
+    suspend fun addMovies(movies: MovieEntity)
+
+    @Query("SELECT * FROM movies_table WHERE id = :movieId")
+    suspend fun getMovies(movieId: Int): MovieEntity
```
what if the movie doesn't exist I think it will return null

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/datasource/local/ImageLocalDataSource.kt:
```diff
@@ -0,0 +1,8 @@
+package com.repository.datasource.local
+
+import com.repository.entity.ImageEntity
+
+interface ImageLocalDataSource {
+    suspend fun addImage(images: ImageEntity)
+    suspend fun getImage(): ImageEntity
+}
```
We will not use this at all

mohammed-akkad committed:
update movie details local data source

mohammed-akkad committed:
refactor movie details local data source
